File: sdk/js/PicarioXPO.RenderAPI.V2/XpoUrlGenerator/GeneralKeys.ts
class GeneralKeys {
    private entityName = "fn";
    private width = "width";
    private height = "height";
    private backgroundColor = "bgcolor";
    private designCaching = "p.dc";
    private resizeMethod = "mode";
    private textureRepeat = "p.r";
    private outputQuality = "quality";
    private imageType = "format";
    private outputType = "ot";
    private sceneThumbnailObjectNumber = "p.objectthumb";
    private highlightObject = "p.highlight";
    private caching = "Cache";
    private coords = "p.coords";
    private watermark = "watermark";
    private frame = "p.frame";
    private renderMode = "p.mode";

    queryStringFormat = "&{0}={1}";

    appendRequest(stringBuilder: string, request: XpoUrlRequest) {

        stringBuilder = stringBuilder.concat("?1=1")
                                     .concat(this.getQueryStringValue(this.width, request.getWidth()))
                                     .concat(this.getQueryStringValue(this.backgroundColor, request.getBackgroundColor()))
                                     .concat(this.getQueryStringValue(this.caching, this.getCachingMethod(request)))
                                     .concat(this.getQueryStringValue(this.height, request.getHeight()))
                                     .concat(this.getQueryStringValue(this.designCaching, request.getDesignCaching()))
                                     .concat(this.getQueryStringValue(this.resizeMethod, this.getResizeMethod(request)))
                                     .concat(this.getQueryStringValue(this.textureRepeat, this.getRepeatMethod(request)))
                                     .concat(this.getQueryStringValue(this.outputQuality, request.getOutputQuality()))
                                     .concat(this.getQueryStringValue(this.imageType, this.getFormat(request.getImageType())))
                                     .concat(this.getQueryStringValue(this.sceneThumbnailObjectNumber, this.getStringValue(request.getSceneThumbnailObjectNumber())))
                                     .concat(this.getQueryStringValue(this.highlightObject, request.getHighlightObject()))
                                     .concat(this.getQueryStringValue(this.watermark, request.getWatermarkImage()))
                                     .concat(this.getQueryStringValue(this.frame, request.getFrame()))
                                     .concat(this.getQueryStringValue(this.renderMode, request.getSceneRenderMode()))
                                     .appendDictionary(request.customParameters);

        if (request.urlType == UrlGeneratorModule.UrlTypes.Coords) {
            stringBuilder = stringBuilder.concat(this.getQueryStringValue(this.coords, true));
        }

        return stringBuilder;
    }

    private getQueryStringValue(key: string, value: any) {
        if (!value)
            return "";

        return this.queryStringFormat.format(key, value);
    }

    private getCachingMethod(request: XpoUrlRequest){
        return request.getCaching() ? "Default" : "No";
    }

    private getRepeatMethod(request: XpoUrlRequest) {
        if (request.getResizeMethod() == UrlGeneratorModule.XpoUrlResizeMethods.Repeat && request.getFileType() == UrlGeneratorModule.XpoUrlFileTypes.Design) {
            return "1";
        }

        return "";
    }

    private getFormat(imageType: UrlGeneratorModule.XpoUrlImageTypes) {
        switch (imageType) {
            case UrlGeneratorModule.XpoUrlImageTypes.Bmp:
                return "bmp";
            case UrlGeneratorModule.XpoUrlImageTypes.Jpg:
                return "jpg";
            case UrlGeneratorModule.XpoUrlImageTypes.Png:
                return "png";
            default:
                return "jpg";
        }
    }

    private getResizeMethod(request: XpoUrlRequest) {
        switch (request.getResizeMethod()) {
            case UrlGeneratorModule.XpoUrlResizeMethods.KeepAspectMax:
                return "max";
            case UrlGeneratorModule.XpoUrlResizeMethods.Crop:
                return "crop";
            case UrlGeneratorModule.XpoUrlResizeMethods.Stretch:
                return "stretch";
            case UrlGeneratorModule.XpoUrlResizeMethods.Canvas:
                return "canvas";
        }

        return "";
    }

    private getStringValue(value: any) {
        return value != null ? value.toString() : "";
    }
}

File: sdk/c#/PicarioXPO.RenderAPI.V2/GeneralKeys.cs
using System.Text;
using PicarioXPO.RenderAPI.V2.Extensions;

namespace PicarioXPO.RenderAPI.V2
{
    internal static class GeneralKeys
    {
        public const string EntityName = "fn";
        public const string Width = "width";
        public const string Height = "height";
        public const string BackgroundColor = "bgcolor";
        public const string DesignCaching = "p.dc";
        public const string ResizeMethod = "mode";
        public const string TextureRepeat = "p.r";
        public const string OutputQuality = "quality";
        public const string ImageType = "format";
        public const string OutputType = "ot";
        public const string SceneThumbnailObjectNumber = "p.objectthumb";
        public const string HighlightObject = "p.highlight";
        public const string Caching = "Cache";
        public const string Coords = "p.coords";
        public const string Watermark = "watermark";
        public const string Frame = "p.frame";
        public const string RenderMode = "p.mode";
        public const string FastRender = "p.fastrender";

        public static StringBuilder AppendRequest(this StringBuilder stringBuilder, XpoUrlRequest request)
        {
            stringBuilder
                .Append("?1=1")
                .Append(Width, (request.Width == 0) ? -1 : request.Width) // Safety (0 is the default value)
                .Append(BackgroundColor, request.BackgroundColor)
                .Append(Caching, GetCachingMethod(request))
                .Append(Height, request.Height)
                .Append(DesignCaching, request.DesignCaching)
                .Append(ResizeMethod, GetResizeMethod(request))
                .Append(TextureRepeat, GetRepeatMethod(request))
                .Append(OutputQuality, request.OutputQuality)
                .Append(ImageType, GetFormat(request.ImageType))
                .Append(SceneThumbnailObjectNumber, request.SceneThumbnailObjectNumber)
                .Append(HighlightObject, request.HighlightObject)
                .Append(Watermark, request.WatermarkImage)
                .Append(Frame, request.Frame)
                .Append(RenderMode, request.SceneRenderMode)
                .Append(FastRender, request.FastRender)
                .AppendDictionary(request.CustomParameters);

            if (request is XpoCoordinatesUrlRequest)
            {
                stringBuilder.Append(Coords, true);
            }
                
            return stringBuilder;
        }

        private static string GetResizeMethod(XpoUrlRequest request)
        {
            switch (request.ResizeMethod)
            {
                case XpoUrlResizeMethods.KeepAspectMax:
                    return "max";
                case XpoUrlResizeMethods.Crop:
                    return "crop";
                case XpoUrlResizeMethods.Stretch:
                    return "stretch";
                case XpoUrlResizeMethods.Canvas:
                    return "canvas";
            }

            return "";
        }


        private static string GetFormat(XpoUrlImageTypes imageType)
        {
            switch (imageType)
            {
                case XpoUrlImageTypes.Bmp:
                    return "bmp";
                case XpoUrlImageTypes.Jpg:
                    return "jpg";
                case XpoUrlImageTypes.Png:
                    return "png";
                default:
                    return "jpg";
            }
        }

        private static string GetRepeatMethod(XpoUrlRequest request)
        {
            if (request.ResizeMethod == XpoUrlResizeMethods.Repeat && request.FileType == XpoUrlFileTypes.Design)
            {
                return "1";
            }
        
            return "";
        }

        private static string GetCachingMethod(XpoUrlRequest request)
        {
            return request.Caching ? "Default" : "No";
        }
    }
}